app/src/main/java/com/eveningoutpost/dexdrip/Models/Tomato.java

```diff
@@ -139,15 +158,15 @@ static void AreWeDone() {
         s_recviedEnoughData = true;
         
         boolean checksum_ok = JoH.LibreCrc(data);
-        Log.e(TAG, "We have all the data that we need checksum_ok = " + checksum_ok + HexDump.dumpHexString(data));
+        Log.e(TAG, "We have all the data that we need " + s_acumulatedSize + " checksum_ok = " + checksum_ok + HexDump.dumpHexString(data));
         
         if(!checksum_ok) {
-            return;
+            throw new RuntimeException(CHECKSUM_FAILED);
         }
         
         PersistentStore.setString("Tomatobattery", Integer.toString(s_full_data[13]));
-        PersistentStore.setString("TomatoHArdware",HexDump.toHexString(s_full_data,14,2));
-        PersistentStore.setString("TomatoFirmware",HexDump.toHexString(s_full_data,16,2));
+        PersistentStore.setString("TomatoHArdware",HexDump.toHexString(s_full_data,16,2));
```

Comment:
order of firmware/hardware swapped but this has already been discussed, but just drawing attention to the change.